src/Dreamberd/Compilation/Preprocessing.hs

```diff
@@ -69,3 +76,16 @@ importFile :: String -> ([String], String) -> IO (Either String ([String], [AstN
 importFile lib (imports, code) = case parseDreamberd (File lib code) of
     Right ast -> preprocessNodes ast lib imports
     Left err -> return $ Left ("Error while importing " ++ lib ++ ": " ++ err)
+
+preprocessFunctions :: [AstNode] -> [AstNode]
+preprocessFunctions a = concatMap hoistFunction functions ++ rest
+  where
+    (functions, rest) = partition isFunction a
+
+isFunction :: AstNode -> Bool
+isFunction Function{} = True
+isFunction _ = False
```

Comment:
You could declare this inside the preprocessFunctions as it is not used anywhere else

Reply:
I need pattern matching,using a lambda here would require a case

Reply:
You can define functions in the where of another function like in the parseChar of Parser.hs